pilot/pkg/model/destination_rule.go

```diff
@@ -38,55 +39,63 @@ import (
 func (ps *PushContext) mergeDestinationRule(p *processedDestRules, destRuleConfig config.Config, exportToMap map[visibility.Instance]bool) {
 	rule := destRuleConfig.Spec.(*networking.DestinationRule)
 	resolvedHost := ResolveShortnameToFQDN(rule.Host, destRuleConfig.Meta)
-
-	if mdr, exists := p.destRule[resolvedHost]; exists {
+	newDr := true
+	if mdrList, exists := p.destRule[resolvedHost]; exists {
 		// Deep copy destination rule, to prevent mutate it later when merge with a new one.
 		// This can happen when there are more than one destination rule of same host in one namespace.
-		copied := mdr.DeepCopy()
-		p.destRule[resolvedHost] = &copied
-		mergedRule := copied.Spec.(*networking.DestinationRule)
-		existingSubset := map[string]struct{}{}
-		for _, subset := range mergedRule.Subsets {
-			existingSubset[subset.Name] = struct{}{}
-		}
-		// we have an another destination rule for same host.
-		// concatenate both of them -- essentially add subsets from one to other.
-		// Note: we only add the subsets and do not overwrite anything else like exportTo or top level
-		// traffic policies if they already exist
-		for _, subset := range rule.Subsets {
-			if _, ok := existingSubset[subset.Name]; !ok {
-				// if not duplicated, append
-				mergedRule.Subsets = append(mergedRule.Subsets, subset)
-			} else {
-				// duplicate subset
-				ps.AddMetric(DuplicatedSubsets, string(resolvedHost), "",
-					fmt.Sprintf("Duplicate subset %s found while merging destination rules for %s",
-						subset.Name, string(resolvedHost)))
+		for i, mdr := range mdrList {
+			copied := mdr.DeepCopy()
+			p.destRule[resolvedHost][i] = &copied
+			mergedRule := copied.Spec.(*networking.DestinationRule)
+			if !reflect.DeepEqual(mergedRule.GetWorkloadSelector().GetMatchLabels(), rule.GetWorkloadSelector().GetMatchLabels()) {
```

Comment:
this is the only major change in the method. remaining green lines are due to the right shift due to the new for loop

Reply:
nit: you can compare maps directly instead of reflect.DeepEquals to imrprove performance.